src/ber/mod.rs

```diff
@@ -0,0 +1,367 @@
+                start - 1
+            } else {
+                start
+            }
+        };
+
+        let length = Length::new(4 - start as u64);
+
+        length.serialize(buffer);
+        buffer.write(&ns[start..]);
+    }
+}
+
+impl Deserialize for u64 {
+    fn deserialize(bytes: &mut &[u8]) -> Result<Self, DeserializeError> {
+        bytes.tag(INTEGER)?;
+        let length = Length::deserialize(bytes)?;
+        Ok(bytes.uint(length)? as u64)
+    }
+}
+
+impl Serialize for u64 {
+    fn serialize(&self, buffer: &mut dyn VecExt) {
+        (*self as i64).serialize(buffer);
+    }
+}
+
+pub struct Null;
+
+impl Deserialize for Null {
+    fn deserialize(buffer: &mut &[u8]) -> Result<Self, DeserializeError> {
+        buffer.tag(NULL)?;
+        let length = Length::deserialize(buffer)?;
+        let _ = buffer.slice(length)?;
+
+        Ok(Null)
+    }
+}
+
+impl Serialize for Null {
+    fn serialize(&self, buffer: &mut dyn VecExt) {
+        NULL.serialize(buffer);
+        Length::new(0).serialize(buffer);
+    }
+}
+
+impl<T: Serialize> Serialize for Vec<T> {
+    fn serialize(&self, buffer: &mut dyn VecExt) {
+        SEQUENCE.serialize(buffer);
+
+        util::serialize_sequence(buffer, |buffer| {
+            for item in self {
+                item.serialize(buffer);
+            }
+        });
+    }
+}
+
+impl<T: Deserialize> Deserialize for Vec<T> {
+    fn deserialize(buffer: &mut &[u8]) -> Result<Self, DeserializeError> {
+        buffer.tag(SEQUENCE)?;
+        let length = Length::deserialize(buffer)?;
+        let slice = &mut buffer.slice(length)?;
+
+        let mut vec = Vec::new();
+        while !slice.is_empty() {
+            vec.push(T::deserialize(slice)?);
+        }
+
+        Ok(vec)
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Set<T>(Vec<T>);
+
+impl<T> From<Vec<T>> for Set<T> {
+    fn from(v: Vec<T>) -> Self {
+        Self(v)
+    }
+}
+
+impl<T: Serialize> Serialize for Set<T> {
+    fn serialize(&self, buffer: &mut dyn VecExt) {
+        SET.serialize(buffer);
+        util::serialize_sequence(buffer, |buffer| {
+            for item in &self.0 {
+                item.serialize(buffer);
+            }
+        });
+    }
+}
+
+impl<T: Deserialize> Deserialize for Set<T> {
+    fn deserialize(buffer: &mut &[u8]) -> Result<Self, DeserializeError> {
+        buffer.tag(SET)?;
+        let length = Length::deserialize(buffer)?;
+        let slice = &mut buffer.slice(length)?;
+
+        let mut vec = Vec::new();
+        while !slice.is_empty() {
+            vec.push(T::deserialize(slice)?);
+        }
+
+        Ok(Self(vec))
+    }
+}
+
+/// LDAP defines a BOOLEAN `true` to be `0xFF` and any other value `false`
+impl Serialize for bool {
+    fn serialize(&self, buffer: &mut dyn VecExt) {
+        BOOL.serialize(buffer);
+        Length::new(1).serialize(buffer);
+
+        match self {
+            true => buffer.write_byte(0xFF),
```

Comment:
clippy will complain about this and tell you to use a normal `if` / `else` instead

Reply:
I actually don't seem to get a warning about this -- I know I have in the past. Regardless, I think the match is more readable than the if else, so I'll just allow the lint in the future if necessary